### PR TITLE
DD-31: Mandate update activity

### PR DESCRIPTION
**Pull request by awg-github**
## Overview

 1. A "Send mandate update notification to the contact?" checkbox is added to the bottom of the "Edit Direct Debit Mandate Record" modal (edit mandate directly) and "New Direct Debit Mandate Record".
2. Adding a postprocessing in the extension so when any change is made to an existing mandate or when a new mandate is used for an existing recurring contribution, a "Direct Debit Mandate Update" activity is generated

dd-31 mandate update activity

### CODE REVIEW

File: CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php [hunk 1]
```diff
@@ -26,10 +26,34 @@ class CRM_ManualDirectDebit_Hook_PostProcess_Contribution_DirectDebitMandate {
...
+   */
+  public function setCurrentContactId($contactId) {
+    $this->currentContactId = $contactId;
+  }
+
+  /**
+   *  Sets id of current mandate
+   */
+  public function setCurrentMandateId() {
+    if (isset($this->form->getVar('_submitValues')['mandateId']) && !empty($this->form->getVar('_submitValues')['mandateId'])) {
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php]: Please re-write like this to improve readability:

```php
$mandateId = $this->form->getVar('_submitValues')['mandateId'];

if (isset($mandateId) && !empty($mandateId)) {
  $this->mandateId = $mandateId;
}
```

File: CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php [hunk 2]
```diff
@@ -54,25 +78,40 @@ public function checkPaymentOptionToCreateMandate() {
...
+  }
+
   /**
    * Changes mandate for recurring contribution
    *
    * @return bool
    */
-  public function changeMandateForRecurringContribution() {
+  public function checkChangingMandateContribution() {
+    if (!isset($this->form->getVar('_submitValues')['recurrId']) || empty($this->form->getVar('_submitValues')['recurrId'])) {
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php]: Same here, we can extract obtaining the value and assign it to a variable to improve readability.

Eg.
```php
$recurrID = $this->form->getVar('_submitValues')['recurrId'];

if (!isset($recuurrId) || empty($recurrId)) {
  ...
}
```

File: CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php [hunk 3]
```diff
@@ -84,10 +123,43 @@ private function redirectToContributionTab() {
...
+    if (!isset($this->form->getVar('_submitValues')['send_mandate_update_notification_to_the_contact']) || empty($this->form->getVar('_submitValues')['send_mandate_update_notification_to_the_contact'])) {
+      return FALSE;
+    }
+
+    $valueOfSendMailCheckbox = $this->form->getVar('_submitValues')['send_mandate_update_notification_to_the_contact'];
+    $isSendMailCheckboxTurnedOn = $valueOfSendMailCheckbox == 1;
+
+    if ($isSendMailCheckboxTurnedOn) {
+
+//      TODO: Sending mail process will be implemented in DD-30
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php]: We really shouldn't merge into develop things that are half done. To be able to finish this, I think we should base this branch (DD-31) off the DD-30 branch, and treat DD-30 as a feature branch or epic. If we do this, the merge of this branch would go into DD-30 branch, and not develop, and we'd merge DD-30 into develop once everything is stable.

Exmple:
```
|
* Merge DD-30 to develop
|\
| * Merge DD-31 to DD-30
| |\
| | |
| | * DD-31
| | | 
| |/
| * DD-30
| | 
|/
| 
| Develop
```

CC @guanhuan

File: CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php [hunk 4]
```diff
@@ -84,10 +123,43 @@ private function redirectToContributionTab() {
...
         'selectedChild' => 'contribute',
       ])
     ));
   }
 
+  /**
+   * Sends mail if appropriate checkbox is checked on
+   */
+  public function checkMailNotification() {
+    if (!isset($this->form->getVar('_submitValues')['send_mandate_update_notification_to_the_contact']) || empty($this->form->getVar('_submitValues')['send_mandate_update_notification_to_the_contact'])) {
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Hook/PostProcess/Contribution/DirectDebitMandate.php]: Another opportunity to improve readability.

**Comment by awg-github** [comment]
Yes, we are waiting  for DD-30  to be merged into develop.

**Pull request merged**